include/revng/Support/AccessCounter.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <cstdint>
+
+#include "revng/Support/Assert.h"
+
+namespace revng {
+
+/**
+ * A AccessCounter is optimized std::stack<bool> that can contain up to 8
+ * elements.
+ **/
+class AccessCounter {
+private:
+  std::uint8_t Counter = 0;
+
+public:
+  bool operator==(const AccessCounter &Other) const {
+    return Counter == Other.Counter;
+  }
+
+  bool operator!=(const AccessCounter &Other) const {
+    return Counter == Other.Counter;
+  }
+  void clear() { Counter &= ~0x1; }
+  void access() { Counter |= 0x1; }
+  void push() {
+    revng_assert(Counter < 128, "More than 8 pushes have been performed");
```

Comment:
that would allow you to push into the stack 9 times is the first time you did not actually accessed the element. It works for a single access counter, but should there be multiple access counters, some accessed and some not, you would get the assert triggering depending on which one you ask

Reply:
How is `Counter < 128` better? It's unclear to me.

Also, we cannot check for `pop`, so it's still best effort.

Reply:
ah wait, somehow i thought the semantic was different. mhh this is very annoying because i will not be able to know how many times a model has been "pushed" by just looking at one of those counters. Not sure we care toh

Reply:
The assertion I suggested does not catch all potential problems but catches those that would actually imply data loss.
If you shift left by one positions, but the most significant bit is `0`, you will get it back when you shift right.
If you shift left by one positions, but the most significant bit is `1`, you would not get it back, but we can detect it.
It's not the best, but it's decent.

Reply:
yeah, i guess i will do count leading zero, it's the same